pkg/git/service.go

```diff
@@ -177,17 +177,61 @@ func (s *Service) SetGitConfig(userData *gitprovider.GitUser) error {
 		}
 	}
 
+	if err := setSigningConfig(cfg, providerConfig); err != nil {
+		return err
+	}
+
 	var buf bytes.Buffer
 	_, err = cfg.WriteTo(&buf)
 	if err != nil {
 		return err
 	}
 
-	err = os.WriteFile(gitConfigFileName, buf.Bytes(), 0644)
+	return os.WriteFile(gitConfigFileName, buf.Bytes(), 0644)
+}
+
+func setSigningConfig(cfg *ini.File, providerConfig *gitprovider.GitProviderConfig) error {
+	if providerConfig == nil || providerConfig.SigningMethod == nil || providerConfig.SigningKey == nil {
+		return nil
+	}
+
+	if !cfg.HasSection("user") {
+		_, err := cfg.NewSection("user")
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err := cfg.Section("user").NewKey("signingkey", *providerConfig.SigningKey)
 	if err != nil {
 		return err
 	}
 
+	if !cfg.HasSection("commit") {
+		_, err := cfg.NewSection("commit")
+		if err != nil {
+			return err
+		}
+	}
+
+	switch *providerConfig.SigningMethod {
+	case gitprovider.SigningMethodGPG:
```

Comment:
How exactly did you test GPG signing?

Reply:
@Tpuljak you will need to specify what problem are you facing. I also faced some problems while testing
you can visit here for troubleshooting
https://docs.gitlab.com/ee/user/project/repository/signed_commits/gpg.html#troubleshooting

Reply:
@bryans-go you're supposed to tell us how to test this 😄 

Please let us know how exactly you tested this and proof that it actually works.
The issue I'm having is that, inside the project, I can't sign with GPG because the key is missing.

Reply:
@Tpuljak 
we will need to do some changes here ```func generateSshConfigEntry(profileId, workspaceId, projectName, knownHostsPath string) (string, error) {
	daytonaPath, err := os.Executable()
	if err != nil {
		return "", err
	}

	tab := "\t"
	projectHostname := GetProjectHostname(profileId, workspaceId, projectName)

	config := fmt.Sprintf("Host %s\n"+
		tab+"User daytona\n"+
		tab+"StrictHostKeyChecking no\n"+
		tab+"UserKnownHostsFile %s\n"+
		tab+"ProxyCommand \"%s\" ssh-proxy %s %s %s\n"+
		tab+"ForwardAgent yes\n"+
		tab+"StreamLocalBindUnlink yes\n"+
		tab+"RemoteForward /home/daytona/.gnupg/S.gpg-agent:/run/user/1000/gnupg/S.gpg-agent.extra\n\n",   // the 1000 is hardcoded we can make it dynamic
		projectHostname, knownHostsPath, daytonaPath, profileId, workspaceId, projectName)

	return config, nil
}```

also after that 

we need to export the public gpg key ```gpg --armor --export 44C255D0CE19BFF3``` // gpg key id submitted by user. copy the full output
then ```daytona ssh```
then ```gpg --import```
paste the output here and then ```ctrl+d```    

this steps can be reduced to   // if we want to code it 
1. ```ssh -R /home/daytona/.gnupg/S.gpg-agent:/run/user/1000/gnupg/S.gpg-agent.extra -o StreamLocalBindUnlink=yes user@remote```

2.Ensure the following directory on the remote exists
if no then  ``` mkdir -p /home/daytona/.gnupg```

Import the public GPG key on the remote (if not done yet):

```gpg --export <key-id> | ssh user@remote 'gpg --import'``` // one line command will need to edit if we want to go this way by automatically adding public gpg in the project
Now we can do the gpg signing flawlessly

Reply:
this requires (OpenSSH 6.7+) and also  see this https://wiki.gnupg.org/AgentForwarding
reference https://superuser.com/questions/161973/how-can-i-forward-a-gpg-key-via-ssh-agent

Reply:
I will commit changes after proper discussion for gpg signing setup

Reply:
sometimes we might need to add this to our bashrc or zshrc ```export GPG_TTY=$(tty)``` for entering password.
also might need to switch between gpg2 and gpg by setting gpg.program = gpg2 in the gitconfig of repo

Reply:
@bryans-go all this seems reasonable and should be automated by Daytona if the user selects GPG signing.

> sometimes we might need to add this to our bashrc or zshrc export GPG_TTY=$(tty) for entering password

Regarding this, you can add that to the users shell as well

> also might need to switch between gpg2 and gpg by setting gpg.program = gpg2 in the gitconfig of repo

I think we can assume `gpg` and then the user, that uses gpg2, will switch themselves.

Regarding the path where we forward the sock, it seems to me that you're mixing `/run/user/1000` and `/home/user`. Let's forward to `/home/user` where the `user` can be fetched by executing `ssh HOST echo '$USER'`. You can find an example of how we fetch the home folder of a project [here](https://github.com/daytonaio/daytona/blob/main/internal/util/path.go#L14). 

Please update the PR with the changes and I'll put the PR in draft until you do so.

In the future, please make sure to test every change and feature your PR addresses and, if not clear, provide steps for us to test it as well.

Reply:
/home/daytona/.gnupg/S.gpg-agent:/run/user/1000/gnupg/S.gpg-agent.extra     it is like this <remote socket><local socket>    am i correct

Reply:
![image](https://github.com/user-attachments/assets/343d883d-b1bf-48a5-86c2-8dfe687b9e50)

Reply:
thats why I use /home/daytona

Reply:
But judging from the resources you provided, the agent should be forwarded to `$HOME/gnupg/S.gpg-agent.extra` inside the project.

Reply:
So `$HOST_HOME/gnupg/S.gpg-agent.extra:$PROJECT_HOME/gnupg/S.gpg-agent.extra`

Reply:
hey which resource said this ?
<remote socket><local extra gpg socket>  is the only correct way that have worked for me

Reply:
https://superuser.com/a/884602

Seems to be here but if `/run/...` only works for you, that's fine as well. You can do it that way.

Reply:
OK i will try that way also .
 but

```
➜  ~ gpgconf --list-dir agent-extra-socket  // local machine
/run/user/1000/gnupg/S.gpg-agent.extra
➜  ~ 
```
https://wiki.gnupg.org/AgentForwarding  this is the official guide

Reply:
we will need to get both socket on local ```gpgconf --list-dir agent-extra-socket``` and on remote ```gpgconf --list-dir agent-socket``` and then set them up

Reply:
Ah okay, then you need to execute that on the users machine and use the output of the command as the forward.

```
gpgconf --list-dir agent-extra-socket
/Users/tomapuljak/.gnupg/S.gpg-agent.extra
```

This is what I get.

Reply:
yes exactly and please I want to know that how will get output of this gpgconf --list-dir agent-extra-socket
/Users/tomapuljak/.gnupg/S.gpg-agent.extra   on remote(dtn ssh)

Reply:
Just as we get `$HOME` [here](https://github.com/daytonaio/daytona/blob/main/internal/util/path.go#L14).

You will execute the command with `ssh` and get the output.

Reply:
OK . thanks for the help

Reply:
@Tpuljak I have added functions in the projected file as you said but I'm unsure about how we can make a if statement here /daytona/cmd/daytona/config/ssh_file.go to check if the git provider selected for project creation have selected signingkeyGPG as a option.
Kindly please suggest a approach or basic idea to start

Reply:
You can use the `apiClient.GitProviderAPI.GetGitProviderForUrl` method to get the git provider config and check if `gpg` is set as the signing method.

Reply:
apiClient.GitProviderAPI.GetGitProviderForUrl(ctx, encodedUrl).Execute()  what will be the encoded url and ctx

Reply:
The URL is the encoded `Repository.Url` of the project you are entering and `ctx` can be `context.Background()`.

Reply:
Please suggest a way to get git provide config because ssh_file.go doesn't let me import apiclient.    Without using apiclient what others things we can do to get git provider config

Reply:
Even after thinking so much and surfing through codebase I'm not able to find a good way to do this

Reply:
Done !